app/src/main/java/com/github/se/polyfit/ui/utils/Authentication.kt

```diff
@@ -4,59 +4,85 @@ import android.app.Activity
 import android.content.Context
 import android.content.Intent
 import android.util.Log
+import androidx.activity.ComponentActivity
 import androidx.activity.result.ActivityResultLauncher
-import androidx.lifecycle.ViewModel
 import co.yml.charts.common.extensions.isNotNull
 import com.firebase.ui.auth.AuthUI
+import com.firebase.ui.auth.FirebaseAuthUIActivityResultContract
 import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult
 import com.github.se.polyfit.data.remote.firebase.UserFirebaseRepository
 import com.github.se.polyfit.model.data.User
 import com.google.android.gms.auth.api.signin.GoogleSignIn
 import com.google.android.gms.auth.api.signin.GoogleSignInAccount
-import dagger.hilt.android.lifecycle.HiltViewModel
-import javax.inject.Inject
-import kotlinx.coroutines.runBlocking
+import com.google.firebase.auth.FirebaseAuth
 
-interface Authentication {
-  fun signIn()
+class Authentication(
+    activity: ComponentActivity,
+    private val user: User,
+    private val userFirebaseRepository: UserFirebaseRepository,
+    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
+    private var context: Context = activity.applicationContext,
+) {
 
-  fun onSignInResult(result: FirebaseAuthUIAuthenticationResult, callback: (Boolean) -> Unit)
+  private var callback: (() -> Unit) = { Log.e("Authentication", "Callback not set") }
+  private var signInLauncher: ActivityResultLauncher<Intent>? = null
+  // This value is used to be sure the request is received in the test.
+  private var isAnswered = false
 
-  fun setSignInLauncher(launcher: ActivityResultLauncher<Intent>)
-}
+  init {
+    initLaunch(activity)
+  }
 
-@HiltViewModel
-class AuthenticationCloud
-@Inject
-constructor(
-    private val context: Context,
-    private val user: User,
-    private val userFirebaseRepository: UserFirebaseRepository
-) : ViewModel(), Authentication {
-  private lateinit var signInLauncher: ActivityResultLauncher<Intent>
+  private fun initLaunch(activity: ComponentActivity) {
+    if (auth.currentUser != null) {
+      isAnswered = true
+      setUserInfo(GoogleSignIn.getLastSignedInAccount(context))
+      return
+    }
+    // This init is launched when the user is not signed in
+    signInLauncher =
+        activity.registerForActivityResult(FirebaseAuthUIActivityResultContract()) { res ->
+          onSignInResult(res) { if (it) (callback)() }
+        }
+    val providers = arrayListOf(AuthUI.IdpConfig.GoogleBuilder().build())
+
+    val signInIntent =
+        AuthUI.getInstance().createSignInIntentBuilder().setAvailableProviders(providers).build()
+
+    signInLauncher!!.launch(signInIntent)
+  }
+
+  fun setCallback(callback: () -> Unit, justToHave2ArguementHilt: Int) {
+    this.callback = callback
+  }
```

Comment:
I need to add a useless argument otherwise Hilt will be confuse with another function

Reply:
This function has to be named like this ?